music_assistant/server/providers/filesystem_smb/__init__.py

```diff
@@ -186,7 +186,13 @@ async def mount(self) -> None:
 
         if platform.system() == "Darwin":
             password_str = f":{password}" if password else ""
-            mount_cmd = f'mount -t smbfs "//{username}:{password_str}@{server}/{share}{subfolder}" "{self.base_path}"'  # noqa: E501
+            mount_cmd = [
+                "mount",
+                "-t",
+                "smbfs",
+                f"//{username}:{password_str}@{server}/{share}{subfolder}",
```

Comment:
You're still going to have the escaping issue on this line, aren't you?

Reply:
Primarily this is about avoiding subshells.

If you pass a `str` to `check_output` it runs `asyncio.create_subprocess_shell`. This is like passing shell=True to the `subprocess` module - rather than creating your target program directly you invoke `/bin/sh -c 'your shell commands here'`. Here, we are subject to shell parsing rules. And that means you can inject a password of `' || rm -rf / || true #` and /bin/sh will return an exit code of 0 and delete your OS. User controlled variables getting near /bin/sh is one of the most common problems you see with modem/router vulnerabilities.

If you pass `list[str]`, it actually invokes `create_subprocess_exec`. Here there is no `/bin/sh`. The target program is invoked directly. There is no interpreting or parsing. `arg[0]` in python becomes `argv[0]` in the target process. It is `/bin/sh` that creates the security issue.

`shlex.quote` was mentioned in one of the tickets. It is way to armor a variable so that when a shell parses it, it arrives at the original variable we had on our side (i.e. to protect it from shell expansion or injection). If there is no /bin/sh, nothing will remove that armor before `mount` is invoked. The target process can't actually can't cope with the armored version.

See also https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.subprocess_shell vs https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.subprocess_exec - the quoting advice only applies to create_subprocess_shell.

So, this PR will stop me from being able to own you by using a dodgey username or password. As a side effect, more symbols in passwords will work. I'd expect `'` to work.

You are correct that we may need to additionally urlencode this URI. For example, an '@' in a password might be a problem. I think you are supposed to use url encoding rules on them (so '@' becomes '%40'). I can't test that, so i haven't done it here.

Reply:
Cool - thanks for taking the time to level up my understanding of how python treats the various subprocess commands as well!